### PR TITLE
Add an new command to tune the score based on a scala file.

### PR DESCRIPTION
**Pull request by sraduvictor**
This command supports only scala files for 8 notes.

I tested it by creating a new score, tuning it and checking that
the tuning as correct.

### CODE REVIEW

File: libmscore/scale.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+//---------------------------------------------------------
+//   getNextLine
+//---------------------------------------------------------
+
+QString Scale::getNextLine(QTextStream& in)
+      {
+      while (!in.atEnd()) {
+            QString line = in.readLine();
+            if (line[0] == '!') {
```
Comment (lasconic) [inline libmscore/scale.cpp]: for one line if, you don't need {}

File: libmscore/scale.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                  continue;
+            }
+            return line;
+      }
```
Comment (lasconic) [inline libmscore/scale.cpp]: the end } needs to be aligned with the previous statement.

File: mscore/scoreview.cpp [hunk 3]
```diff
@@ -5991,5 +6006,14 @@ void ScoreView::updateContinuousPanel()
             update();
       }
 
+//---------------------------------------------------------
+//   cmdAttachScale
```
Comment (lasconic) [inline mscore/scoreview.cpp]: cmdTuneToScale

File: mscore/scoreview.cpp [hunk 4]
```diff
@@ -3061,6 +3062,20 @@ void ScoreView::cmd(const QAction* a)
...
             }
+      else if (cmd == "tune_to_scale") {
+            QString name = mscore->getScalaFilename();
+            if (!name.isEmpty()) {
+                  _score->startCmd();
+                  Scale s;
+                  if (!s.loadScale(name)) {
+                        QMessageBox::critical(this,
+                                              tr("MuseScore: Load Scala File"), MScore::lastError);
+                  } else {
```
Comment (lasconic) [inline mscore/scoreview.cpp]: `else {` should be on a new line and } should be indented.

**Comment by sraduvictor** [comment]
I made the style changes, according to https://musescore.org/en/developers-handbook/musescore-coding-rules

**Comment by sraduvictor** [comment]
Changed a foreach into a for

**Comment by sraduvictor** [comment]
In the next commit I'll attach the scale to a score and set the tuning in NPlayEvent from playNote in rendermidi.cpp, I'll probably need to show this scale somewhere, maybe in Synthetizer in the Tuning panel.
And the commit after it I'll save the scale to file so we don't need to set it again.

**Comment by orchestralmusic-net** [comment]
Thank you for your effort, it would be really nice to be able to set tuning in MuseScore, but as far as I can see, you have implemented only Scala files with 8 notes, which are rare and really non-standard. Any meaningful historical tuning needs to use Scala files with 12 notes. You can always edit your 8-notes Scala file and change it to 12-notes, but not vice-versa. So to obtain compatibility, MuseScore surely needs to read (at least) 12-notes files.
Here is a link with some 4500 Scala files, containing many historical tunings (any many others):
http://www.huygens-fokker.org/scala/downloads.html#scales
Some of the most important historical tunings are Meantone (Quarter-comma), Werckmeister III and Valotti-Young. These tunings are also used in historical informed baroque or classical concerts. It would be nice to use these three as test-cases for automatic testing.
If there is interest, I can collect a little collection of the most important historical tunings in Scala format to be shipped with MuseScore (or downloaded as extension). These all would be public domain (no problem with licence).
If someone is interested, here is a link to my site featuring some analysis of the most important historical tunings (I use it as a comparison with my own invented tunings, which are the first three in this link):
http://www.pure-tuning.net/scripts/temperaments-comparison.php?lang=eng

**Comment by mgavioli** [comment]
Using scales/temperaments with 8 notes per octave (actually seven, I presume, from F to B) is of rather limited usefulness, as there is very little music using only those notes.

12 notes per octave (Eb - G# or similar) would be the absolute minimum, better 17 (Gb - A#) or 21 (Fb - B#). But in fact the easiest would be to support the full complement of 35 notes per octave known to MuseScore (Fbb - B##), filling the notes not supported by a given scale/temperament using the enharmonies.

In fact, I fail to see any reason for not supporting it.

**Comment by mgavioli** [comment]
Additionally, I believe the possibility for the user to edit the scale(s) or temperament(s) to be important.

This is linked to a basic limitation of Scala format, which lists the notes in pitch order, without noting what is what. For instance, in the Zalino temperament, all #'s are below their b's enharmonies, while in the Pythagorean scale, all #'s are above their b's enharmonies. So, the Scala format would list C# before Db for the former, but Db before C# for the latter; and any algorithm reading them would attributes the wrong Tonal Pitch Classes to one or to the other.

So:

1) the user has to be able to correct the scales and temperaments (and, why not, creates his own)

2) Scala format cannot be used in the score data, but only used in import as a source of info.

**Comment by lasconic** [comment]
Good to see you around @mgavioli 
It's a start :) Also I believe Byzantine music, @sraduvictor's use case has 8 notes per octave.

This PR was meant to duplicate the feature of @mgavioli's plugin and help @sraduvictor to know the MuseScore code a bit. The feature in this PR will be less relevant if he manage to implement the scale import in rendermidi.cpp and make the temperament a property of the score. You can see the progress here https://github.com/musescore/MuseScore/pull/2405 and I will close this PR.

My understanding is that @sraduvictor knowledge regarding temperament is limited to Byzantine Music. Mine is limited... so I have some questions for you @mgavioli and @orchestralmusic-net. Please post your finding here https://musescore.org/en/node/99666 (I'm still making a summary of your posts here)

**Pull request closed without merging**